### PR TITLE
fix: crash during fetchOrAssert on NSObjectContext [WPB-5393]

### PR DESCRIPTION
**Pull request by markusfassbender**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-5393" title="WPB-5393" target="_blank">BugWPB-5393</a>  [iOS] Crash during fetchOrAssert on NSObjectContext
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

We found crashes stack traces in Xcode organizer that crashes the app hard, whenever an object can not be fetched from Core Data

### Solutions

Use `assertionFailure` to crash only in debug over `fatalError` that crashes also in production.

#### How to Test

See ticket, unfortunately it does not crash always.

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: wire-ios-data-model/Source/ManagedObjectContext/NSManagedObjectContext+FetchRequest.swift [hunk 1]
```diff
@@ -27,7 +27,8 @@ public extension NSManagedObjectContext {
             return result
         } catch let error {
             WireLogger.localStorage.error("CoreData: Error in fetching request : \(request),  \(error.localizedDescription)")
-            fatal("Error in fetching \(error.localizedDescription)")
+            assertionFailure("Error in fetching \(error.localizedDescription)")
+            return []
```
Comment (markusfassbender) [inline wire-ios-data-model/Source/ManagedObjectContext/NSManagedObjectContext+FetchRequest.swift:31]: By replacing the custom `fatal` crash, we then miss:
`ZMAssertionDump_NSString("Swift assertion", "\(file)", Int32(line), message)`

Is this dump used for anything, do we need this additionally to `WireLogger.localStorage.error` in line 29?

Comment (markusfassbender) [reply wire-ios-data-model/Source/ManagedObjectContext/NSManagedObjectContext+FetchRequest.swift:31]: @johnxnguyen @typfel @netbe @David-Henner @agisilaos @caldrian @theswiftway ?

Comment (netbe) [reply wire-ios-data-model/Source/ManagedObjectContext/NSManagedObjectContext+FetchRequest.swift:31]: I would say add the WireLogger too. I am not sure about the dump

Comment (johnxnguyen) [reply wire-ios-data-model/Source/ManagedObjectContext/NSManagedObjectContext+FetchRequest.swift:31]: I think it's fine to leave the dump out. On another note, do you know what the output of `"\(request)"` looks like? Will it give us anything useful to know which request failed?

Comment (markusfassbender) [reply wire-ios-data-model/Source/ManagedObjectContext/NSManagedObjectContext+FetchRequest.swift:31]: @netbe sorry how do you mean this, there is the line right above:
`WireLogger.localStorage.error("CoreData: Error in fetching request : \(request),  \(error.localizedDescription)")`

Comment (markusfassbender) [reply wire-ios-data-model/Source/ManagedObjectContext/NSManagedObjectContext+FetchRequest.swift:31]: @johnxnguyen I've printed the statement it formats this way:
> "CoreData: Error in fetching request : <NSFetchRequest: 0x60000292b560> (entity: Label; predicate: (type == 1); sortDescriptors: ((null)); limit: 2; type: NSManagedObjectResultType; ),  A database key is required to migrate."

Comment (johnxnguyen) [reply wire-ios-data-model/Source/ManagedObjectContext/NSManagedObjectContext+FetchRequest.swift:31]: ok looks good to me

**Review by johnxnguyen — APPROVED** [review]

**Pull request merged**